File: Medium/2026-02-09-1382-Balance-a-Binary-Search-Tree/solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public TreeNode balanceBST(TreeNode root) {
        List<Integer> nodes = new ArrayList<>();
        inorderTraversal(root, nodes);
        return buildBalancedBST(nodes, 0, nodes.size() - 1);
    }

    private void inorderTraversal(TreeNode node, List<Integer> nodes) {
        if (node == null) return;
        inorderTraversal(node.left, nodes);
        nodes.add(node.val);
        inorderTraversal(node.right, nodes);
    }

    private TreeNode buildBalancedBST(List<Integer> nodes, int start, int end) {
        if (start > end) return null;
        int mid = start + (end - start) / 2;
        TreeNode root = new TreeNode(nodes.get(mid));
        root.left = buildBalancedBST(nodes, start, mid - 1);
        root.right = buildBalancedBST(nodes, mid + 1, end);
        return root;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

File: Medium/2026-02-09-1382-Balance-a-Binary-Search-Tree/solution.js
class TreeNode {
    constructor(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

var balanceBST = function(root) {
    const nodes = [];

    const inorderTraversal = (node) => {
        if (!node) return;
        inorderTraversal(node.left);
        nodes.push(node.val);
        inorderTraversal(node.right);
    };

    inorderTraversal(root);

    const buildBalancedBST = (start, end) => {
        if (start > end) return null;
        const mid = Math.floor((start + end) / 2);
        const node = new TreeNode(nodes[mid]);
        node.left = buildBalancedBST(start, mid - 1);
        node.right = buildBalancedBST(mid + 1, end);
        return node;
    };

    return buildBalancedBST(0, nodes.length - 1);
};